### PR TITLE
feat(otto):mint timestamp & summon

### PR DESCRIPTION
**Pull request by d0c0q**

### CODE REVIEW

File: contracts/Otto.sol [hunk 1]
```diff
@@ -235,3 +235,69 @@ contract Otto is
 
     function _authorizeUpgrade(address) internal override onlyAdmin {}
 }
+
+contract OttoV2 is Otto, IOttoV2 {
```
Comment (otterclamking) [inline contracts/Otto.sol:239]: Feel no need to create v2. We can just update origin contract.

File: contracts/Otto.sol [hunk 2]
```diff
@@ -235,3 +235,69 @@ contract Otto is
...
+    uint256 public incubationPeriod;
+
+    function setIncubationPeriod(uint256 incubationPeriod_)
```
Comment (otterclamking) [inline contracts/Otto.sol:242]: setSummonPeriod

File: contracts/Otto.sol [hunk 3]
```diff
@@ -235,3 +235,69 @@ contract Otto is
...
+            infos[startTokenId + i] = OttoInfo({
+                name: '',
+                description: '',
+                birthday: 0,
+                traits: 0,
+                values: 0,
+                attributes: 0,
+                attributeBonuses: 0,
+                flags: 0,
+                __reserved: [block.timestamp, 0, 0, 0, 0, 0]
```
Comment (otterclamking) [inline contracts/Otto.sol:274]: Can't we edit the struct field name?

File: contracts/OttoV2.sol [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+        pure
+        returns (uint256 traits_)
+    {
+        traits_ = 0;
+        for (uint16 i = 0; i < 16; i++) {
+            traits_ = (traits_ << 16) | arr_[15 - i];
+        }
+    }
+
+    function U256toU16(uint256 traits_)
```
Comment (otterclamking) [inline contracts/OttoV2.sol:369]: We can extract this method to library and use `using` on type

File: contracts/OttoV2.sol [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+    function totalMintable() external view virtual override returns (uint256) {
+        return collectionSize - totalSupply();
+    }
+
+    function maxBatch() external view virtual override returns (uint256) {
+        return maxBatchSize;
+    }
+
+    function minted(uint256 tokenId_)
```
Comment (otterclamking) [inline contracts/OttoV2.sol:280]: We could just call this `exists`

File: contracts/OttoV2.sol [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+        require(
+            candidateIndex < candidates[tokenId_].length,
+            'invalid candidate index'
+        );
+        uint8[32] memory flags_ = U256toU8(infos[tokenId_].flags);
+        flags_[0] = uint8(PortalStatus.SUMMONED);
+        infos[tokenId_].flags = U8toU256(flags_);
+        infos[tokenId_].traits = candidates[tokenId_][candidateIndex];
+        infos[tokenId_].birthday = birthday_;
+        emit OttoSummoned(
```
Comment (otterclamking) [inline contracts/OttoV2.sol:245]: Let's clear candidates array after summoned

File: contracts/OttoV2.sol [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        string memory name_,
+        string memory description_,
+        uint256 tokenId_,
+        uint256 birthday_,
+        uint256 traits_,
+        uint256 values_,
+        uint256 attributes_,
+        uint256 attributeBonuses_,
+        uint256 flags_
+    ) external virtual onlyManager validOttoId(tokenId_) {
```
Comment (otterclamking) [inline contracts/OttoV2.sol:179]: feel it's easier to separate the setter like:
`setName`
`setDescription`
`setBirthday`
`setTraits`...

File: contracts/OttoV2.sol [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+        string name;
+        string description;
+        uint256 birthday;
+        uint256 traits; // uint8 [...]
+        uint256 values; // uint32 [level, experiences, hungerValue, friendship, ...reserved]
+        // int16[] [STR, DEF, DEX, INT, LUK, CON, CUTE, BRS, ...reserved]
+        uint256 attributes; // can be changed by level up
+        uint256 attributeBonuses; // from traits & wearable
+        uint256 flags; // bool [portalStatus, isLegendary, ...reserved]
+        uint256 mintTimestamp;
```
Comment (otterclamking) [inline contracts/OttoV2.sol:76]: I prefer mintAt

Comment (otterclamking) [inline contracts/OttoV2.sol:76]: let's also track `summonAt`

File: contracts/OttoV2.sol [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+        view
+        virtual
+        override
+        validOttoId(tokenId_)
+        returns (bool)
+    {
+        return U256toU8(infos[tokenId_].flags)[1] != 0;
+    }
+
+    function canSummonTimestamp(uint256 tokenId_)
```
Comment (otterclamking) [inline contracts/OttoV2.sol:323]: canSummonAt

File: contracts/OttoV2.sol (hunk 5 above)
Comment (d0c0q) [reply contracts/OttoV2.sol]: https://github.com/OtterClam/otter-contracts/pull/16/commits/de95a0cba2ac6df1d874b983ba28a4d50d81a553

File: contracts/OttoV2.sol (hunk 6 above)
Comment (d0c0q) [reply contracts/OttoV2.sol:249]: https://github.com/OtterClam/otter-contracts/pull/16/commits/de95a0cba2ac6df1d874b983ba28a4d50d81a553

File: contracts/OttoV2.sol (hunk 8 above)
Comment (d0c0q) [reply contracts/OttoV2.sol]: https://github.com/OtterClam/otter-contracts/pull/16/commits/de95a0cba2ac6df1d874b983ba28a4d50d81a553

Comment (d0c0q) [reply contracts/OttoV2.sol]: https://github.com/OtterClam/otter-contracts/pull/16/commits/de95a0cba2ac6df1d874b983ba28a4d50d81a553

File: contracts/OttoV2.sol (hunk 9 above)
Comment (d0c0q) [reply contracts/OttoV2.sol]: https://github.com/OtterClam/otter-contracts/pull/16/commits/de95a0cba2ac6df1d874b983ba28a4d50d81a553

File: contracts/OttoV2.sol (hunk 7 above)
Comment (d0c0q) [reply contracts/OttoV2.sol]: https://github.com/OtterClam/otter-contracts/pull/16/commits/7b25d78fe2bd0f38ce7de9058d63ca51fb26ae93

File: contracts/OttoV2.sol (hunk 4 above)
Comment (d0c0q) [reply contracts/OttoV2.sol]: I found that struct in mapping is upgradeable, so I change the needed variable type to reduce the complexity of encoding uint256.

see https://github.com/OtterClam/otter-contracts/pull/16/commits/7b25d78fe2bd0f38ce7de9058d63ca51fb26ae93

File: contracts/OttoV2.sol [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+    }
+
+    function setTraits(uint256 tokenId_, uint256 traits_)
+        external
+        virtual
+        override
+        onlyAdmin
+        validOttoId(tokenId_)
+    {
+        infos[tokenId_].traits = traits_;
```
Comment (otterclamking) [inline contracts/OttoV2.sol:235]: need a `SetTraits` event

File: contracts/OttoV2.sol [hunk 11]
```diff
@@ -143,6 +144,33 @@ contract OttoV2 is
...
+    {
+        for (uint256 i = 0; i < tokendIds_.length; i++) {
+            require(_exists(tokendIds_[i]), 'invalid tokenId');
+            infos[tokendIds_[i]].canOpenAt = ts_;
+        }
+    }
+
+    function setTraits(uint256 tokenId_, uint256 traits_)
+        external
+        onlyAdmin
```
Comment (otterclamking) [inline contracts/OttoV2.sol:168]: feel manager can `setTraits` ?

**Review by otterclamking — APPROVED** [review]

File: contracts/OttoV2.sol (hunk 11 above)
Comment (d0c0q) [reply contracts/OttoV2.sol]: https://github.com/OtterClam/otter-contracts/pull/16/commits/536e60084b42ffb25f060335da04771cae68be67

**Pull request merged**